tests/drivers/flash_nop_device/src/main.c

```diff
@@ -0,0 +1,198 @@
+			off, r_val32, PATTERN8TO32BIT(FLASH_NOP_DEVICE_ERASE_VALUE));
+	}
+}
+
+static void test_erase(void)
+{
+	int rc;
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET +
+			 FLASH_NOP_DEVICE_ERASE_UNIT,
+			 FLASH_NOP_DEVICE_ERASE_UNIT);
+	zassert_ok(rc, "flash_erase should succeed");
+}
+
+static void test_out_of_bounds(void)
+{
+	int rc;
+	uint8_t data[8] = {0};
+
+	rc = flash_write(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_write(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_write(flash_dev, TEST_NOP_FLASH_END,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_write(flash_dev, TEST_NOP_FLASH_END - 4,
+				 data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET -
+			 FLASH_NOP_DEVICE_ERASE_UNIT,
+			 FLASH_NOP_DEVICE_ERASE_UNIT);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, TEST_NOP_FLASH_END,
+			 FLASH_NOP_DEVICE_ERASE_UNIT);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET -
+			 FLASH_NOP_DEVICE_ERASE_UNIT * 2,
+			 FLASH_NOP_DEVICE_ERASE_UNIT * 2);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, TEST_NOP_FLASH_END -
+			 FLASH_NOP_DEVICE_ERASE_UNIT,
+			 FLASH_NOP_DEVICE_ERASE_UNIT * 2);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET - 4,
+				 data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, TEST_NOP_FLASH_END,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_read(flash_dev, TEST_NOP_FLASH_END - 4, data, 8);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+}
+
+static void test_align(void)
+{
+#if FLASH_NOP_DEVICE_PROG_UNIT > 1
+	int rc;
+	uint8_t data[4] = {0};
+
+	rc = flash_write(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET + 1,
+				 data, 4);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_write(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET,
+				 data, 3);
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET + 1,
+			 FLASH_NOP_DEVICE_ERASE_UNIT);
+
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
+
+	rc = flash_erase(flash_dev, FLASH_NOP_DEVICE_BASE_OFFSET,
+			 FLASH_NOP_DEVICE_ERASE_UNIT + 1);
+
+	zassert_equal(-EINVAL, rc, "Unexpected error code (%d)", rc);
```

Comment:
Here we should check for FLASH_NOP_DEVICE_ERASE_UNIT

Reply:
We are

Reply:
We check for the prog_unit, but since flash_simulator has a hidden dependence between these two we may leave it as it is